File: Striver Important Interview Questions/Day 4 - Arrays Part 4/longest-substring-without-repeating-characters.cpp

// https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0;
        int n = s.size();
        if(n == 0) return ans;

        int left = 0, right = 0;
        vector<int> arr(256, -1);

        while(right < n){
            if(arr[s[right]] != -1) left = max(left, arr[s[right]]+1);
            arr[s[right]] = right;
            ans = max(ans, (right-left+1));
            right++;
        }
        
        return ans;
    }
};

File: Neetcode-150-js/Sliding window/longest-substring-without-repeating-characters.js

// https://leetcode.com/problems/longest-substring-without-repeating-characters/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let len = s.length;
    let ans = 0;
    if(len === 0) return ans;

    let left = 0, right = 0;
    let map = {};
    while(right < len){
        if(map[s[right]] !== undefined) left = Math.max(left, map[s[right]]+1);
        map[s[right]] = right;
        ans = Math.max(ans, right-left+1);
        right++;
    }
    return ans;
};